doc/man3/tst_tmpdir.3
.\"
.\" $Id: tst_tmpdir.3,v 1.1 2000/07/27 16:59:03 alaffin Exp $
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of version 2 of the GNU General Public License as
.\" published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it would be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" Further, this software is distributed without any warranty that it is
.\" free of the rightful claim of any third person regarding infringement
.\" or the like.  Any license provided herein, whether implied or
.\" otherwise, applies only to this software file.  Patent licenses, if
.\" any, provided herein do not apply to combinations of this program with
.\" other software, or any other product whatsoever.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
.\" Mountain View, CA  94043, or:
.\"
.\" http://www.sgi.com
.\"
.\" For further information regarding this notice, see:
.\"
.\" http://oss.sgi.com/projects/GenInfo/NoticeExplan/
.\"
.TH TST_TMPDIR 3 07/25/2000 "Linux Test Project"
.SH NAME
tst_tmpdir \- create a unique testing directory and make it current.
.br
tst_rmdir \- remove the directory created by \fBtst_tmpdir\fR.
.SH SYNOPSIS
\fBvoid tst_tmpdir()
.P
void tst_rmdir()
.P
extern char *TESTDIR;\fR
.SH DESCRIPTION
The \fBtst_tmpdir()\fR function uses the first three characters of the
\fBTCID\fR global variable as the prefix in forming a unique directory name
(via \fBtempnam\fR(3S)).  The directory is then created and made the current
working directory.
.P
If \fBtst_tmpdir()\fR cannot form a unique directory name, create the
directory, or \fBchdir\fR to the directory, it uses \fBtst_brk()\fR to issue
"BROK" messages for all test cases.  It then exits via \fBtst_exit()\fR.
Because \fBtst_tmpdir()\fR exits in the event of a problem, a test must call
it \fBbefore\fR performing any operations that would require running a
cleanup routine.
.P
The \fBtst_rmdir()\fR function recursively removes the directory created by
\fBtst_tmpdir()\fR.  This function should be used \fBonly\fR as a companion
to \fBtst_tmpdir()\fR and should be called immediately prior to the test
exiting via \fBtst_exit()\fR.
.P
\fBtst_rmdir()\fR uses the \fBsystem\fR(3S) library routine (which in turn
calls \fBfork\fR(2)), so tests that use it \fBcannot\fR treat SIGCHLD as an
unexpected signal.
.P
Users may gain access to the name of the temporary directory by declaring the
external character pointer \fBTESTDIR\fR.
.SH DIAGNOSTICS
The \fBtst_rmdir()\fR function will check the \fBTESTDIR\fR global variable
to ensure that the user is not attempting to remove the root directory or
some unspecified directories with a "*" parameter.  All error/warning
messages are delivered through \fBtst_resm()\fR.
.SH "SEE ALSO"
fork(2), system(3S), tst_res(3), tmpnam(3S).



ChangeLog
LTP-20100831

commit dd59d49d75598d163f1b4599067a9f0dcb70e6c4
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Tue Aug 31 23:09:49 2010 -0700

    The cacheflush system call does not return EINVAL in the latest Linux
    kernel. see the link:
    A related patch about the cacheflush function:
    http://lkml.org/lkml/2009/4/9/203
    But Ralf had refused this patch:
    http://www.spinics.net/lists/linux-man/msg00906.html
    cacheflush01 testcase checks if the return value is EINVAL when the
    cache argument is not one of ICACHE, BCACHE or DCACHE. So it will fail
    in all boards.
    
    In this modification, checking return value SUCCESS will be added,
    instead of checking EINVAL
    
    Signed-off-by: Feng Gao <feng.gao@windriver.com>

:100644 100644 cce0b70... d0a4b1f... M	testcases/kernel/syscalls/cacheflush/cacheflush01.c

commit ebea45b92a1e30e4480f25291ef4c1ef2eeaa4bf
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Tue Aug 31 17:14:21 2010 -0700

    Get rid of unnecessary style divergence.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 0ebeca9... b08a2d9... M	testcases/commands/at/at_allow01
:100755 100755 6bca177... 9f9f433... M	testcases/commands/at/at_deny01

commit 08640ca184c1fc1e24e966c3d806a722f3ca2154
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 20:53:18 2010 -0700

    Add more assertions.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:000000 100644 0000000... 2c590b1... A	testcases/open_posix_testsuite/conformance/interfaces/access/assertions.xml

commit 3cdde2f86af90b3fd4e2b1468a12adfd2a9523f0
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 20:52:52 2010 -0700

    Note which timer failed in output.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 ac043b0... 8e149ea... M	testcases/open_posix_testsuite/conformance/behavior/timers/1-1.c

commit 207f47c4c9b4783c1a18be32431f5a1b319b9e5a
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 19:34:34 2010 -0700

    Update some assertions for raise(2) to SUSv4 reqs.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 38d9542... aa47b7c... M	testcases/open_posix_testsuite/conformance/interfaces/raise/assertions.xml

commit 9c6d2b4cf4f7c166813ddee280ab2bf128eb521c
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 03:36:51 2010 -0700

    Fix bad syntax that I thought I hadn't committed.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 f731d6c... 0ebeca9... M	testcases/commands/at/at_allow01

commit 414747ced26d943a22a3f767deb7f5b4658adb6e
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 03:32:18 2010 -0700

    Forgot the sleeps.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 9b10030... 6d5eb7f... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-2.c
:100644 100644 a3be617... 60e5241... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-3.c
:100644 100644 34e3afd... abc3279... M	testcases/open_posix_testsuite/conformance/interfaces/pthread_attr_setstack/2-1.c
:100755 100755 01984c5... 536d407... M	testcases/open_posix_testsuite/scripts/generate-makefiles.sh

commit 3222fbecbfb2924a6b901a8e2048c70e1a993018
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:22:07 2010 -0700

    Remove bogus comment.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 c7d81b5... 72a1fcd... M	testcases/open_posix_testsuite/conformance/behavior/timers/2-1.c

commit 7187f0921382bf375cd111b8a480a74ce5786945
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:18:23 2010 -0700

    Repeat robustness work done to 1-1 for 1-2 and 1-3.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 a9d104c... 9b10030... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-2.c
:100644 100644 af23ebe... a3be617... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-3.c

commit 6c2750a55eaa30fe708c78087688c0d468bfb807
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:12:07 2010 -0700

    Make test more robust.
    
    1. Catch fork() == -1 errors.
    2. Sleep so wait(2) should always succeed.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 1efd520... f46ee8f... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-1.c

commit 873f6e5b28a6a23367a7117aad436378e56ad09a
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:42:53 2010 -0700

    Fix ltp-aiodio.sh script so it execs aio-stress.
    
    Singed-off-by: Jacky Malcles <jacky.malcles@bull.net>

:100755 100755 ed2bce4... ab54c54... M	testscripts/ltp-aiodio.sh

commit 4fd92183833eed128bb6420ece7a6dd45dcd8ec5
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:40:51 2010 -0700

    I've prepared a patch for at_allow01 that essentially duplicates the
    fixes that were checked in for at_deny01. This fixes the script to
    actually run the test when the at utility in stead of returning TCONF,
    correcting test environment, and let the script run normally without
    changing the password. See commit
    a589d12b0b49287d10f43b1a1fb72950819588eb for more details.
    
    Signed-off-by: Henry Yei <hyei@mvista.com>

:100755 100755 cf5b97c... f731d6c... M	testcases/commands/at/at_allow01

commit 217d9b778d275bcbaf5a9fa9e1a5ac2716521cad
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:39:12 2010 -0700

    Chain cmds together to avoid intermediate errors.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 b94bd8f... 6bca177... M	testcases/commands/at/at_deny01

commit 9eeea7939bd73b295ae376dc045e5fbb984d0f38
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Wed Aug 25 20:40:23 2010 -0700

    Fix error message (was hardcoded to SIGALARM).
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>


lib/tst_tmpdir.c
/**********************************************************
 * Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it would be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * Further, this software is distributed without any warranty that it is
 * free of the rightful claim of any third person regarding infringement
 * or the like.  Any license provided herein, whether implied or
 * otherwise, applies only to this software file.  Patent licenses, if
 * any, provided herein do not apply to combinations of this program with
 * other software, or any other product whatsoever.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
 * Mountain View, CA  94043, or:
 *
 * http://www.sgi.com
 *
 * For further information regarding this notice, see:
 *
 * http://oss.sgi.com/projects/GenInfo/NoticeExplan/
 *********************************************************/

/**********************************************************
 *
 *	 OS Testing - Silicon Graphics, Inc.
 *
 *	 FUNCTION NAME	  : tst_tmpdir, tst_rmdir
 *
 *	 FUNCTION TITLE	 : Create/remove a testing temp dir
 *
 *	 SYNOPSIS:
 *		void tst_tmpdir();
 *		void tst_rmdir();
 *
 *	 AUTHOR		 : Dave Fenner
 *
 *	 INITIAL RELEASE	: UNICOS 8.0
 *
 *	 DESCRIPTION
 *		tst_tmpdir() is used to create a unique, temporary testing
 *		directory, and make it the current working directory.
 *		tst_rmdir() is used to remove the directory created by
 *		tst_tmpdir().
 *
 *	 RETURN VALUE
 *		Neither tst_tmpdir() or tst_rmdir() has a return value.
 *
 *********************************************************/

#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <assert.h>
#include <errno.h>
#include <libgen.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "test.h"
#include "rmobj.h"
#include "ltp_priv.h"
#include "lapi/futex.h"

/*
 * Define some useful macros.
 */
#define DIR_MODE	(S_IRWXU|S_IRWXG|S_IRWXO)

#ifndef PATH_MAX
#ifdef MAXPATHLEN
#define PATH_MAX	MAXPATHLEN
#else
#define PATH_MAX	1024
#endif
#endif

/*
 * Define global variables.
 */
extern char *TCID;		/* defined/initialized in main() */
static char *TESTDIR = NULL;	/* the directory created */

static char test_start_work_dir[PATH_MAX];

/* lib/tst_checkpoint.c */
extern futex_t *tst_futexes;

int tst_tmpdir_created(void)
{
	return TESTDIR != NULL;
}

char *tst_get_tmpdir(void)
{
	/* Smack the user for calling things out of order. */
	if (TESTDIR == NULL)
		tst_brkm(TBROK, NULL, "you must call tst_tmpdir() first");
	return strdup(TESTDIR);
}

const char *tst_get_startwd(void)
{
	return test_start_work_dir;
}

void tst_tmpdir(void)
{
	char template[PATH_MAX];
	char *env_tmpdir;
	char *errmsg, *c;

	/*
	 * Create a template for the temporary directory.  Use the
	 * environment variable TMPDIR if it is available, otherwise
	 * use our default TEMPDIR.
	 */
	env_tmpdir = getenv("TMPDIR");
	if (env_tmpdir) {
		c = strchr(env_tmpdir, '/');
		/*
		 * Now we force environment variable TMPDIR to be an absolute
		 * pathname, which dose not make much sense, but it will
		 * greatly simplify code in tst_rmdir().
		 */
		if (c != env_tmpdir) {
			tst_brkm(TBROK, NULL, "You must specify an absolute "
				 "pathname for environment variable TMPDIR");
		}
		snprintf(template, PATH_MAX, "%s/%.3sXXXXXX", env_tmpdir, TCID);
	} else {
		snprintf(template, PATH_MAX, "%s/%.3sXXXXXX", TEMPDIR, TCID);
	}

	/* Make the temporary directory in one shot using mkdtemp. */
	if (mkdtemp(template) == NULL)
		tst_brkm(TBROK | TERRNO, NULL,
			 "%s: mkdtemp(%s) failed", __func__, template);
	if ((TESTDIR = strdup(template)) == NULL)
		tst_brkm(TBROK | TERRNO, NULL,
			 "%s: strdup(%s) failed", __func__, template);

	if (chown(TESTDIR, -1, getgid()) == -1)
		tst_brkm(TBROK | TERRNO, NULL,
			 "chown(%s, -1, %d) failed", TESTDIR, getgid());
	if (chmod(TESTDIR, DIR_MODE) == -1)
		tst_brkm(TBROK | TERRNO, NULL,
			 "chmod(%s, %#o) failed", TESTDIR, DIR_MODE);

	if (getcwd(test_start_work_dir, sizeof(test_start_work_dir)) == NULL) {
		tst_resm(TINFO, "Failed to record test working dir");
		test_start_work_dir[0] = '\0';
	}

	/*
	 * Change to the temporary directory.  If the chdir() fails, issue
	 * TBROK messages for all test cases, attempt to remove the
	 * directory (if it was created), and exit.  If the removal also
	 * fails, also issue a TWARN message.
	 */
	if (chdir(TESTDIR) == -1) {
		tst_resm(TERRNO, "%s: chdir(%s) failed", __func__, TESTDIR);

		/* Try to remove the directory */
		if (rmobj(TESTDIR, &errmsg) == -1) {
			tst_resm(TWARN, "%s: rmobj(%s) failed: %s",
				 __func__, TESTDIR, errmsg);
		}

		tst_exit();
	}
}

void tst_rmdir(void)
{
	char *errmsg;

	/*
	 * Check that TESTDIR is not NULL.
	 */
	if (TESTDIR == NULL) {
		tst_resm(TWARN,
			 "%s: TESTDIR was NULL; no removal attempted",
			 __func__);
		return;
	}

	/*
	 * Unmap the backend file.
	 * This is needed to overcome the NFS "silly rename" feature.
	 */
	if (tst_futexes) {
		msync((void *)tst_futexes, getpagesize(), MS_SYNC);
		munmap((void *)tst_futexes, getpagesize());
	}

	/*
	 * Attempt to remove the "TESTDIR" directory, using rmobj().
	 */
	if (rmobj(TESTDIR, &errmsg) == -1) {
		tst_resm(TWARN, "%s: rmobj(%s) failed: %s",
			 __func__, TESTDIR, errmsg);
	}
}
